File: src/main/resources/info/gianlucacosta/aurora/gradle/tasks/RunJava.js
var shell = WScript.CreateObject("WScript.Shell")
var fileSystemObject = WScript.CreateObject("Scripting.FileSystemObject")


var programFiles64BitPath = "C:/Program Files/Java"
var programFiles32BitPath = "C:/Program Files (x86)/Java" 


var standardJavaDirRegex = /(?:jdk|jre)(\d+)\.(\d+)(?:\.(\d+)(?:_(\d+))?)?(?:-.+)?/

var javaVersionOutpuRegex = /java version "(\d+)\.(\d+)(?:\.(\d+)(?:_(\d+))?)?(?:-.+)?"/

var officialJavaUrl = "https://java.com/download/"

var argumentsPartitioner = "+++"
		


main()


function main() {
	var scriptDir = fileSystemObject.GetParentFolderName(WScript.ScriptFullName)
	var appBaseDirectory = fileSystemObject.GetParentFolderName(scriptDir)


	var scriptArguments = WScript.Arguments

	var requiredJavaVersion = buildVersion(Array(
			scriptArguments(0),
			scriptArguments(1),
			scriptArguments(2),
			scriptArguments(3)
	))

	var mainClass = scriptArguments(4)

	var runWithJavaw = (parseInt(scriptArguments(5)) == 1) 

	var additionalJvmArguments = Array()
	var appArguments = Array()

	var argumentsPartitionerFound = false

	for (i = 6; i < scriptArguments.length; i++)
	{
		var currentScriptArgument = scriptArguments(i)

		if (currentScriptArgument == argumentsPartitioner) {
			argumentsPartitionerFound = true
		} else {
			if (!argumentsPartitionerFound) {
				additionalJvmArguments.push(currentScriptArgument)
			} else {
				appArguments.push(currentScriptArgument)
			}
		}		
	}

	WScript.Echo()
	WScript.Echo()

	WScript.Echo("Additional JVM arguments: " + additionalJvmArguments)
	WScript.Echo("App arguments: " + appArguments)
	
	WScript.Echo()
	WScript.Echo()


	WScript.Echo("REQUIRED JAVA VERSION: " + requiredJavaVersion)

	WScript.Echo()
	WScript.Echo()

	var javaInfo = getJavaInfo(requiredJavaVersion)

	WScript.Echo()
	WScript.Echo()

	if (javaInfo) {
		WScript.Echo("COMPATIBLE JAVA VERSION FOUND: " + javaInfo.version +  " @ " + javaInfo.home)
		WScript.Echo()

		WScript.Echo("Main class: " + mainClass)


		WScript.Echo("App base directory: " + appBaseDirectory)

		var appLibDirectory = appBaseDirectory + "/lib"		

		var classPathArguments = Array(
				"-cp",
				'"' + appLibDirectory + '/*"'				
			)


		var overallJavaArguments = classPathArguments.concat(additionalJvmArguments)
		overallJavaArguments.push(mainClass)
		overallJavaArguments = overallJavaArguments.concat(appArguments)

		var javaCommandLine = getJavaCommandLine(
			javaInfo.home,
			runWithJavaw,
			overallJavaArguments
		)

		WScript.Echo()
		WScript.Echo("Now running the Java app with the following command line:")
		WScript.Echo("---> " + javaCommandLine)
		WScript.Echo()

		try {	
			var appExitCode
			
			if (runWithJavaw) {
				shell.Run(javaCommandLine)
				appExitCode = 0
			} else {
				appExitCode = shell.Run(javaCommandLine, 1, true)
			}
			
			WScript.Quit(appExitCode)
		} catch(ex) {			
			WScript.Echo("Error while running the app! Code: " + ex.number)
			WScript.Quit(1)
		}		
	} else {
		showWarning("No compatible Java version found!\n\nNow opening Java's official download website.")
		shell.Run(officialJavaUrl)
		
		WScript.Quit(1)
	}
}


function getJavaInfo(requiredJavaVersion) {
	var environmentJavaHomeInfo = searchEnvironmentJavaHome(requiredJavaVersion)
	if (environmentJavaHomeInfo) {
		return environmentJavaHomeInfo
	}

	WScript.Echo()
	WScript.Echo()
	
	var programFiles64BitInfo = search64BitProgramFilesDirectory(requiredJavaVersion)
	if (programFiles64BitInfo) {
		return programFiles64BitInfo
	}

	WScript.Echo()
	WScript.Echo()

	var programFiles32BitInfo = search32BitProgramFilesDirectory(requiredJavaVersion)
	if (programFiles32BitInfo) {
		return programFiles32BitInfo
	}

	WScript.Echo()
	WScript.Echo()

	return null
}


function searchEnvironmentJavaHome(requiredJavaVersion) {
	try {
		var environmentJavaHome = shell.ExpandEnvironmentStrings("%JAVA_HOME%")

		if (environmentJavaHome != "%JAVA_HOME%")  {
			WScript.Echo("JAVA_HOME found: " + environmentJavaHome)

			var javaExeFilePath = getJavaExeFilePath(environmentJavaHome)

			if (!fileSystemObject.FileExists(javaExeFilePath)) {
				WScript.Echo("This JAVA_HOME does not actually contain a Java interpreter!")
				return null
			}


			var environmentJavaHomeMatch = standardJavaDirRegex.exec(environmentJavaHome)

			var environmentJavaHomeVersion = null

			if (environmentJavaHomeMatch) {
				environmentJavaHomeVersion = buildVersion(environmentJavaHomeMatch.slice(1))
			} else {
				WScript.Echo("It is not a standard Java directory name - now asking Java itself")
				WScript.Echo()

				var javaCommandLine = getJavaCommandLine(
					environmentJavaHome, 
					false, 
					Array("-version")
				)


				WScript.Echo("Running Java command to get version info:")
				WScript.Echo("---> " + javaCommandLine)

			
				var javaExec = shell.Exec(javaCommandLine)
		
				var javaVersionLine = javaExec.StdErr.ReadLine()			

				if (javaExec.ExitCode == 0) {
					WScript.Echo("Detected Java version line --> " + javaVersionLine)

					var javaVersionOutputMatch = javaVersionOutpuRegex.exec(javaVersionLine)

					if (javaVersionOutputMatch) {
						environmentJavaHomeVersion = buildVersion(javaVersionOutputMatch.slice(1))
						WScript.Echo("Detected Java version--> " + environmentJavaHomeVersion)
					} else {
						WScript.Echo("Cannot detect the Java version via its output line")
					}
				} else {
					WScript.Echo("Execution error while trying to detect Java version")
				}
			}

			if (environmentJavaHomeVersion) {
				WScript.Echo("JAVA_HOME version: " + environmentJavaHomeVersion)

				if (compareVersions(environmentJavaHomeVersion, requiredJavaVersion) >= 0) {				
						return {
							home: environmentJavaHome,
							version: environmentJavaHomeVersion
						}
					} else {
						WScript.Echo("Obsolete Java version in JAVA_HOME")
					}
			}
		} else {
			WScript.Echo("JAVA_HOME not set!")
		}
	} catch(ex) {
		WScript.Echo("An exception occurred: " + ex.description)
	}
}



function getJavaExeFilePath(javaHome) {
	return javaHome + "/bin/java.exe"
}


function getJavawExeFilePath(javaHome) {
	return javaHome + "/bin/javaw.exe"
}


function search64BitProgramFilesDirectory(requiredJavaVersion) {
	WScript.Echo("Now scanning for 64-bit Java in '" + programFiles64BitPath + "'...")

	return scanForTheLatestJavaDirectory(programFiles64BitPath, requiredJavaVersion)
}


function search32BitProgramFilesDirectory(requiredJavaVersion) {
	WScript.Echo("Now scanning for 32-bit Java in '" + programFiles32BitPath + "'...")

	return scanForTheLatestJavaDirectory(programFiles32BitPath, requiredJavaVersion)
}


function scanForTheLatestJavaDirectory(rootDirectoryPath, requiredJavaVersion) {
	try {
		var rootDirectory = fileSystemObject.GetFolder(rootDirectoryPath)	

		var highestJavaInfo = null	

		for (var subFoldersEnum = new Enumerator(rootDirectory.SubFolders); !subFoldersEnum.atEnd(); subFoldersEnum.moveNext()){	
			var subFolder = subFoldersEnum.item()
			var currentMatch = standardJavaDirRegex.exec(subFolder.name)
			
			if (currentMatch) {
				var currentRelativePath = currentMatch[0]
				var currentVersion = buildVersion(currentMatch.slice(1))

				WScript.Echo("Java found ------> " + currentVersion + " @ " + currentRelativePath)

				if (
					!highestJavaInfo || (				
						compareVersions(currentVersion, highestJavaInfo.version) > 0
					)
				) {			
					var currentAbsolutePath = fileSystemObject.GetFolder(rootDirectory + "/" + currentRelativePath).path
					var javaExeFilePath = getJavaExeFilePath(currentAbsolutePath)

					if (fileSystemObject.FileExists(javaExeFilePath)) {
						highestJavaInfo = {
							home: currentAbsolutePath,
							version: currentVersion
						}
					} else {
						WScript.Echo("(Skipping this directory, as it does not contain a Java interpreter)")
						WScript.Echo()
					}
				}
			}
		}
		
		WScript.Echo()

		if (
			highestJavaInfo && 
			(compareVersions(highestJavaInfo.version, requiredJavaVersion) >= 0)
		) {
			return highestJavaInfo
		} else {
			WScript.Echo("No updated Java version found!")

			return null
		}
	} catch(ex) {
		WScript.Echo("An exception occurred: " + ex.description)
	}
}



function getJavaCommandLine(javaHome, useJavaw, arguments) {
	var interpreterFilePath =
		useJavaw ?
			getJavawExeFilePath(javaHome)
			: 
			getJavaExeFilePath(javaHome)

	var argumentsString = arguments.join(" ")

	return '"' + interpreterFilePath + '"' + " " + argumentsString
}


function showWarning(prompt) {
  WScript.Echo(prompt)
  shell.Popup(prompt, 0, "Warning", 0x30)  
}


function buildVersion(versionComponents) {
	var result = Array()

	for (var i = 0; i <= 3; i++) {
		result.push(
			versionComponents[i] ?
				parseInt(versionComponents[i])
				:
				0
		)
	}

	return result
}


function compareVersions(thisVersion, thatVersion) {
	for (var i = 0; i <= 3; i++) {
		var componentDifference = thisVersion[i] - thatVersion[i]
		if (componentDifference != 0) {
			return componentDifference
		}
	}

	return 0
}

File: src/main/resources/info/gianlucacosta/aurora/gradle/tasks/RunJava.py
import re
import os
import sys
import subprocess
import webbrowser


standardJavaDirPattern = re.compile(r"(?:jdk|jre)(\d+)\.(\d+)(?:\.(\d+)(?:_(\d+))?)?(?:-.+)?")

javaVersionOutputPattern = re.compile(r'java version "(\d+)\.(\d+)(?:\.(\d+)(?:_(\d+))?)?(?:-.+)?"')


officialJavaUrl = "https://java.com/download/"

argumentsPartitioner = "+++"



def main():
	scriptDir =  os.path.dirname(os.path.realpath(__file__))
	appBaseDirectory = os.path.dirname(scriptDir)


	requiredJavaVersion = buildVersion(sys.argv[1:5])

	mainClass = sys.argv[5]

	additionalScriptArguments = sys.argv[6:]

	partitionerIndex = additionalScriptArguments.index(argumentsPartitioner)

	additionalJvmArguments = additionalScriptArguments[:partitionerIndex]
	appArguments = additionalScriptArguments[(partitionerIndex + 1):]


	print("")
	print("")

	print("Additional JVM arguments: {}".format(additionalJvmArguments))
	print("App arguments: {}".format(appArguments))
	
	print("")
	print("")


	print("REQUIRED JAVA VERSION: {}".format(requiredJavaVersion))
	print("")
	print("")

	javaInfo = getJavaInfo(requiredJavaVersion)

	print("")
	print("")

	if javaInfo:
		print("COMPATIBLE JAVA VERSION FOUND: {} @ {} ".format(javaInfo["version"], javaInfo["home"]))
		print("")

		print("Main class: {}".format(mainClass))

		print("App base directory: {}".format(appBaseDirectory))

		appLibDirectory = "{}/lib".format(appBaseDirectory)

		classPathArguments = [
				"-cp",
				'"{}/*"'.format(appLibDirectory)
			]

		overallJavaArguments = classPathArguments + additionalJvmArguments
		overallJavaArguments += [mainClass]
		overallJavaArguments += appArguments

		javaCommandLine = getJavaCommandLine(
			javaInfo["home"],			
			overallJavaArguments
		)

		print("")
		print("Now running the Java app with the following command line:")
		print("---> {}".format(javaCommandLine))
		print("")

		try:
			appExitCode = subprocess.call(javaCommandLine, shell=True)

			sys.exit(appExitCode)
		except Exception as ex:
			print("Error while running the app! {}".format(ex))
			sys.exit(1)
		
	else:
		showWarning("No compatible Java version found!\n\nNow opening Java's official download website.")
		webbrowser.open(officialJavaUrl)
		sys.exit(1)



def getJavaInfo(requiredJavaVersion):
	environmentJavaHomeInfo = searchEnvironmentJavaHome(requiredJavaVersion)
	if environmentJavaHomeInfo is not None:
		return environmentJavaHomeInfo

	print("")
	print("")

	systemPathJavaHomeInfo = searchSystemPathJavaHome(requiredJavaVersion)
	if systemPathJavaHomeInfo is not None:
		return systemPathJavaHomeInfo

	return None



def searchEnvironmentJavaHome(requiredJavaVersion):
	try:
		environmentJavaHome = os.environ.get("JAVA_HOME")

		if environmentJavaHome:
			environmentJavaHome = os.path.abspath(os.path.realpath(environmentJavaHome))

			print("JAVA_HOME found: {}".format(environmentJavaHome))

			if os.path.isdir(environmentJavaHome):
				environmentJavaHomeMatch = standardJavaDirPattern.search(environmentJavaHome)

				environmentJavaHomeVersion = None

				if environmentJavaHomeMatch:
					environmentJavaHomeVersion = buildVersion(environmentJavaHomeMatch.group(1, 2, 3, 4))
				else:
					print("It is not a standard Java directory name - now asking Java itself")
					print("")

					environmentJavaHomeVersion = runJavaToGetVersion(environmentJavaHome)


				if environmentJavaHomeVersion:
					print("JAVA_HOME version found: {}".format(environmentJavaHomeVersion))

					if environmentJavaHomeVersion >= requiredJavaVersion:
						return {
							"home": environmentJavaHome,
							"version": environmentJavaHomeVersion
						}
					else:
						print("Obsolete Java version in JAVA_HOME")
					
			else:
				print("JAVA_HOME does not reference an existing directory")
		else:
			print("JAVA_HOME not set!")
	except Exception as ex:
		print("An exception occurred: {}".format(ex))



def runJavaToGetVersion(javaHome):
	javaCommandLine = getJavaCommandLine(
		javaHome,
		["-version"]
	)

	print("Running Java command to get version info:")
	print("---> {}".format(javaCommandLine))

	javaProcess = subprocess.Popen(javaCommandLine, shell=True, stderr=subprocess.PIPE)

	javaVersionLine = javaProcess.stderr.readline()

	print("Detected Java version line --> {}".format(javaVersionLine))

	javaVersionOutputMatch = javaVersionOutputPattern.search(javaVersionLine)

	if javaVersionOutputMatch:
		print("Detected Java version--> {}".format(javaVersionOutputMatch.group(0)))
		return buildVersion(javaVersionOutputMatch.group(1, 2, 3, 4))
	else:
		print("Cannot detect the Java version via its output line")
		return None



def searchSystemPathJavaHome(requiredJavaVersion):
	print("Now searching the system PATH environment variable...")
	print("")

	try:
		javaPath = subprocess.check_output(["which", "java"]).strip()

		if javaPath:
			javaHome = os.path.dirname(
				os.path.dirname(
					javaPath
				)
			)

			print("Java found in PATH! --> {}".format(javaHome))

			javaPathVersion = runJavaToGetVersion(javaHome)

			if javaPathVersion:
				return {
					"home": javaHome,
					"version": javaPathVersion
				}
		else:
			print("Java is not in the system PATH")


	except subprocess.CalledProcessError:
		print("Java is not in the system PATH")



def getJavaCommandLine(javaHomePath, arguments):
	javaExeSubPath =  "/bin/java"

	argumentsString = " ".join(arguments)

	return '"{}{}" {}'.format(javaHomePath, javaExeSubPath, argumentsString)



def showWarning(prompt):	
	print(prompt)

	try:
		subprocess.call(
			'zenity --warning --text="{}"'.format(prompt),
			shell=True,
			stdout=subprocess.PIPE,
			stderr=subprocess.PIPE
		)
	except:
		try:
			subprocess.call(
			r'osascript -e "tell app \"System Events\" to display dialog \"{}\" buttons \"OK\" default button 1 with title \"Warning\" with icon caution"'.format(prompt),
			shell=True,
			stdout=subprocess.PIPE,
			stderr=subprocess.PIPE
		)
		except:
			pass


def buildVersion(versionComponents):
	return tuple([
		int(versionComponent) if versionComponent else 0 for versionComponent in versionComponents
	])


if __name__ == "__main__":
	main()